flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java

```diff
@@ -406,14 +407,40 @@ public void testBoundedBuffer() throws Exception {
 		assertEquals(1, localBufferPool.getNumberOfAvailableMemorySegments());
 	}
 
+	@Test
+	public void testMaxBuffersPerChannelAndAvailability() throws IOException, InterruptedException {
```

Comment:
I would either add another more complicated test case, or expand this one to cover more scenarios (two blocked channels + at least one channel that exceeded limit by more than one buffer):
```
// same setup, but with 3 subpartitions
requestBufferBuilderBlocking(0)
requestBufferBuilderBlocking(1)
assertTrue(...isDone())

requestBufferBuilderBlocking(0)
assertFalse(...isDone())

requestBufferBuilderBlocking(0)
requestBufferBuilderBlocking(2)
requestBufferBuilderBlocking(2)

assertFalse(...isDone())
recycle(1)
assertFalse(...isDone())
recycle(2)
assertFalse(...isDone())
recycle(0)
assertFalse(...isDone())
recycle(0)
assertTrue(...isDone())
(...)
```
?